### PR TITLE
add note regarding forks

### PR DESCRIPTION
**Pull request by jdkandersson**
This PR follows on from this conversation: https://github.com/canonical/operator-workflows/pull/93#issuecomment-1427182448

It proposes that PRs from forks should be redirected at a separate branch if any of the status checks cannot be run (which means that any repos using `operator-workflows` need to have this be done). The branch protection can be updated to decline PRs directly from forks using a similar pattern as the required status checks job. PR to be created.

### CODE REVIEW

**Review by yhaliaw — COMMENTED** [review]
Is there any guidance on naming of these separate branches?
It might be good to set a standard naming convention now.

**Review by yanksyoon — DISMISSED** [review]
LGTM

**Comment by jdkandersson** [comment]
> Is there any guidance on naming of these separate branches?
> It might be good to set a standard naming convention now.

Good point, added

**Review by yhaliaw — APPROVED** [review]
LGTM

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -299,6 +299,14 @@ exposes our team and Canonical to operational risks.
...
 - Bypassing of the rules is disabled.
 - Commits must be signed.
+- PRs from forks should not be merged directly into the default branch if any of
+  the CI/CD cannot be run (e.g., because it requires a secret). This is because
+  there might be issues with the PR that are missed because certain checks did
+  not run. These PRs should be redirected to a separate branch (prefixed with
+  `fork-*`) and then have a PR raised from that branch to the default branch.
+  The PR should link to the original PR from the fork for reference. The branch
+  protection of a repo should not allow branches from forks to be merged
+  directly into the default branch.
```
Comment (mthaddon) [inline CONTRIBUTING.md:309]: How would this work for external contributors?

Comment (nrobinaubertin) [reply CONTRIBUTING.md:309]: That means that we would need to create a `fork-*` branch for them I imagine.  
This will be tedious if there is a lot of external contributors.

Comment (jdkandersson) [reply CONTRIBUTING.md:309]: yes it would require manually creating a branch for any PRs, I'm guessing this can be automated using GitHub actions as external contributions increase. Can you think of any alternatives?

Comment (arturo-seijas) [reply CONTRIBUTING.md:309]: It is likely that will be able to get rid of the limitations. I've filed https://github.com/canonical/charmhub.io/issues/1468. Perhaps it would be worth to take a look ourselves.

Comment (jdkandersson) [reply CONTRIBUTING.md:309]: In some cases, like wordpress, some tests require secrets. If/ when we re-enable `upload-charm-docs`, that also requires secrets for the dry run mode. The guide already indicates that going into a fork branch is only needed if some of the tests can't run. If they can all run, then the fork branch isn't needed

Comment (nrobinaubertin) [reply CONTRIBUTING.md:309]: Instead of having a separate branch for all external contributors, we could have only one that they will target with their PRs. It eases branch management a bit.  

Another idea is that we use another protected branch internally (other than main) that automatically gets commits from the main branch when they pass the tests (it could be named `prod`). This also would be the branch used to publish the charm from and create new versions.  
This would effectively make our CI 2-steps (some tests would perhaps only make sense on main).

Comment (jdkandersson) [reply CONTRIBUTING.md:309]: We probably will never be able to automatically merge into a branch from forks since that opens up an attack vector of getting to run a workflow on the repo from forks without human intervention. At best we can automatically merge into a branch and raise a PR once the PR from the fork has been approved by 2 maintainers on the repository.

Re having a specific branch to target, we would have to keep that branch up to date with the default branch. I'm not sure what value it adds?

Basically, there is no way around requiring human intervention before the full test suite is run where some of the tests require secrets. Otherwise, we leak the secret. Any branch on the repository has permission to access the secrets, so there can't be a path to land code on a branch without human review first

Comment (nrobinaubertin) [reply CONTRIBUTING.md:309]: > We probably will never be able to automatically merge into a branch from forks since that opens up an attack vector of getting to run a workflow on the repo from forks without human intervention. At best we can automatically merge into a branch and raise a PR once the PR from the fork has been approved by 2 maintainers on the repository.

I was not proposing an automatic merge. I thought that PR from forks could target `main`. After being validated and merged to `main`, they still would not be in our `prod` branch who's pulling from `master` when we feel that it's okay.

But I agree that it's not so clean as your proposal since we would mix external changes (maybe not entirely vetted by our full CI) and ours (that are using the full CI).

I think that it's not an issue to do as you propose right now since there is not a lot of external contributors.

Comment (amandahla) [reply CONTRIBUTING.md:309]: I'm not sure if you already consider this: Wouldn't be possible to manually approve to run a workflow in a forked PR and then merge as usual?
Like here:
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

Comment (jdkandersson) [reply CONTRIBUTING.md:309]: That addresses a different issue, even if approved the workflow doesn't get access to the repo secrets so the tests requiring secrets won't pass

Comment (gregory-schiano) [reply CONTRIBUTING.md:309]: I feel like we're twisting things there, which tends to make me think we doing something wrong in our workflow.
How are other opensource project doing to manage forks and maintain secrets ?

Comment (jdkandersson) [reply CONTRIBUTING.md:309]: It isn't unusual for PRs to open source repos to be merged into a different branch first to make modifications. Many open source projects don't require secrets for their tests to run which is why they don't have this problem. We do have those tests - hence this problem

**Comment by weiiwang01** [comment]
Actually, I am not a fan of creating a new branch and PR ourself every time an outside contributor wants to contribute. Creating a new branch and new PR that the outside contributor can not edit might take away outside contributor's agency and discourage the outside contributor.

I think another approach for this problem is to use special comments to trigger actions for PR initiated by outside contributors. You can find an example here: https://github.com/orgs/community/discussions/25389

**Comment by gregory-schiano** [comment]
> Actually, I am not a fan of creating a new branch and PR ourself every time an outside contributor wants to contribute. Creating a new branch and new PR that the outside contributor can not edit might take away outside contributor's agency and discourage the outside contributor.
> 
> I think another approach for this problem is to use special comments to trigger actions for PR initiated by outside contributors. You can find an example here: https://github.com/orgs/community/discussions/25389

So the process would be:
- A contributor forks the repo
- Makes a branch and does a change
- Creates a PR to our main repo
- This PR will run a limited workflow (one that doesn't requires secrets/permissions)
- Someone from the team can the workflow steps which couldn't run from the contributor permissions

Few remarks:
- We'll have to keep the "success checks" enabled, that will lead to workflow concidered as failed when it's triggered by the PR, but will mark the workflow as success when the workflow is triggered by the comment, allowing to merge the PR
- Whenever a change is push to the branch we'll have to comment to re-run the workflow, couldn't this be done by a bot ?

Overall I like this approach, and I've seen it multiple times in public repository.

**Comment by arturo-seijas** [comment]
> > Actually, I am not a fan of creating a new branch and PR ourself every time an outside contributor wants to contribute. Creating a new branch and new PR that the outside contributor can not edit might take away outside contributor's agency and discourage the outside contributor.
> > I think another approach for this problem is to use special comments to trigger actions for PR initiated by outside contributors. You can find an example here: https://github.com/orgs/community/discussions/25389
> 
> So the process would be:
> 
> * A contributor forks the repo
> * Makes a branch and does a change
> * Creates a PR to our main repo
> * This PR will run a limited workflow (one that doesn't requires secrets/permissions)
> * Someone from the team can the workflow steps which couldn't run from the contributor permissions
> 
> Few remarks:
> 
> * We'll have to keep the "success checks" enabled, that will lead to workflow concidered as failed when it's triggered by the PR, but will mark the workflow as success when the workflow is triggered by the comment, allowing to merge the PR
> * Whenever a change is push to the branch we'll have to comment to re-run the workflow, couldn't this be done by a bot ?
> 
> Overall I like this approach, and I've seen it multiple times in public repository.
+1
Sounds like a nice solution.

**Comment by jdkandersson** [comment]
> So the process would be:
> 
> * A contributor forks the repo
> * Makes a branch and does a change
> * Creates a PR to our main repo
> * This PR will run a limited workflow (one that doesn't requires secrets/permissions)
> * Someone from the team can the workflow steps which couldn't run from the contributor permissions
> 
> Few remarks:
> 
> * We'll have to keep the "success checks" enabled, that will lead to workflow concidered as failed when it's triggered by the PR, but will mark the workflow as success when the workflow is triggered by the comment, allowing to merge the PR
> * Whenever a change is push to the branch we'll have to comment to re-run the workflow, couldn't this be done by a bot ?
> 
> Overall I like this approach, and I've seen it multiple times in public repository.

I agree this is a much better approach. I don't think we can ever have a bot that kicks of the workflow runs automatically in response to community contributions since that gives a path to any developer to execute anything they want to. At that point, we may as well just allow everyone to access the secrets on the repo

**Comment by gregory-schiano** [comment]
> > So the process would be:
> > 
> > * A contributor forks the repo
> > * Makes a branch and does a change
> > * Creates a PR to our main repo
> > * This PR will run a limited workflow (one that doesn't requires secrets/permissions)
> > * Someone from the team can the workflow steps which couldn't run from the contributor permissions
> > 
> > Few remarks:
> > 
> > * We'll have to keep the "success checks" enabled, that will lead to workflow concidered as failed when it's triggered by the PR, but will mark the workflow as success when the workflow is triggered by the comment, allowing to merge the PR
> > * Whenever a change is push to the branch we'll have to comment to re-run the workflow, couldn't this be done by a bot ?
> > 
> > Overall I like this approach, and I've seen it multiple times in public repository.
> 
> I agree this is a much better approach. I don't think we can ever have a bot that kicks of the workflow runs automatically in response to community contributions since that gives a path to any developer to execute anything they want to. At that point, we may as well just allow everyone to access the secrets on the repo

Yes indeed, and a manual comment is not that bad and forces us to have a look at the PR before launching the full workflow

**Comment by javierdelapuente** [comment]
I will close this PR as the last message is from 2023. Please reopen it is you think it is still relevant.
